### PR TITLE
Errors for External Service Tasks

### PR DESCRIPTION
**Pull request by philippfromme**
# Changes

* add _Errors_ group for external service tasks
* fix select empty state for global elements

chrome_DBfRs71M3e

### CODE REVIEW

**Comment by philippfromme** [comment]
I found an issue with the edited state after applying the fix to the selects. The empty value is now `empty` instead of an empty string. `isEdited` interprets that as edited. It should be configurable like:

```javascript
function isEdited(node, emptyValue = '') {
  return node && node.value !== emptyValue;
}

isEdited(node, 'empty');
```

**Comment by MaxTru** [comment]
General question / suggestion: 
* Would it not make sense to call the Section `Throw errors`? I think the current label `Errors` does not convey the message, that a user can specify which error shall be thrown based on a expression provided. Also cf. the [docs about this](https://docs.camunda.org/manual/7.15/user-guide/process-engine/external-tasks/#error-event-definitions)

WDYT? @philippfromme @andreasgeier @nikku

**Comment by MaxTru** [comment]
The error is not cleaned up, when we switch away from Implementationtype==external. Is there already a behavior that does this / or do we have an issue to create such a behavior?

**Comment by philippfromme** [comment]
@MaxTru I've copied the labels from the old properties panel:

Camunda_Modeler_KSyTSuugUD

I just noticed that the error label shows the error code if it has been set. That part is actually missing in my implementation. Shall I adjust it?

File: src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js [hunk 1]
```diff
@@ -80,6 +81,7 @@ export default class CamundaPlatformPropertiesProvider {
     const groups = [
       ProcessVariablesGroup(element),
       ImplementationGroup(element),
+      ErrorsGroup(element),
```
Comment (MaxTru) [inline src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js:84]: Should we not add a testcase in CamundaPlatformPropertiesProvider.spec.js ?

File: src/provider/camunda-platform/properties/Error.js [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  };
+
+  const getValue = () => {
+    return errorEventDefinition.get('camunda:expression');
+  };
+
+  return TextField({
+    element: errorEventDefinition,
+    id: idPrefix + '-expression',
+    label: translate('Throw Expression'),
```
Comment (MaxTru) [inline src/provider/camunda-platform/properties/Error.js:94]: Sentence case should be the new standard
```suggestion
    label: translate('Throw expression'),
```

File: src/provider/camunda-platform/properties/Error.js [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+      });
+    });
+
+    return options;
+  };
+
+  return ReferenceSelect({
+    element,
+    id: idPrefix + '-errorRef',
+    label: translate('Global Error Reference'),
```
Comment (MaxTru) [inline src/provider/camunda-platform/properties/Error.js:192]:
```
    label: translate('Global error reference'),
```

File: src/provider/camunda-platform/properties/Error.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    autoFocusEntry: 'errorName',
```
Comment (MaxTru) [inline src/provider/camunda-platform/properties/Error.js:193]: This has no effect for me. E.g., if I uncollapse an error entry (which has a reference), the errorName is not focussed. Also, should it not be the throwExpression, that is automatically selected once I create a new error?

**Comment by MaxTru** [comment]
> @MaxTru I've copied the labels from the old properties panel:
> 
> Camunda_Modeler_KSyTSuugUD
> 
> I just noticed that the error label shows the error code if it has been set. That part is actually missing in my implementation. Shall I adjust it?

Yeah, let's do that. I remember that back then we decided to deliberately include it, after consultants explained us, that the code is the one important attribute about an error (because it is used for correlation)

File: src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js (hunk 1 above)
Comment (philippfromme) [reply src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js:84]: You're right, I just realized, that we test this as well. Will add tests!

File: src/provider/camunda-platform/properties/Error.js (hunk 4 above)
Comment (philippfromme) [reply src/provider/camunda-platform/properties/Error.js:193]: This should actually be `idPrefix + "-errorName"`. Good catch!

File: src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js (hunk 1 above)
Comment (philippfromme) [reply src/provider/camunda-platform/CamundaPlatformPropertiesProvider.js:84]: Done (cf. https://github.com/bpmn-io/bpmn-properties-panel/pull/108/files#diff-3e285c619700ed55ad027c61886b2c397ce0defbe03746dae9debefb028869ec).

File: src/provider/camunda-platform/properties/Error.js (hunk 2 above)
Comment (philippfromme) [reply src/provider/camunda-platform/properties/Error.js]: Done (cf. https://github.com/bpmn-io/bpmn-properties-panel/pull/108/files#diff-e9ae2a1607ddb924b0430102a13f4e6591870d01ce2777ba90d86e146ea9c816R94).

File: src/provider/camunda-platform/properties/Error.js (hunk 3 above)
Comment (philippfromme) [reply src/provider/camunda-platform/properties/Error.js]: Done (cf. https://github.com/bpmn-io/bpmn-properties-panel/pull/108/files#diff-e9ae2a1607ddb924b0430102a13f4e6591870d01ce2777ba90d86e146ea9c816R192).

File: src/provider/camunda-platform/properties/Error.js (hunk 4 above)
Comment (philippfromme) [reply src/provider/camunda-platform/properties/Error.js]: Done (cf. https://github.com/bpmn-io/bpmn-properties-panel/pull/108/files#diff-e9ae2a1607ddb924b0430102a13f4e6591870d01ce2777ba90d86e146ea9c816R193).

**Comment by barmac** [comment]
There's a merge conflict ;(

**Comment by philippfromme** [comment]
> There's a merge conflict ;(

Will resolve. 👍🏻

**Comment by MaxTru** [comment]
> The error is not cleaned up, when we switch away from Implementationtype==external. Is there already a behavior that does this / or do we have an issue to create such a behavior?

I couldn't find such a behavior and created a task in https://github.com/bpmn-io/bpmn-properties-panel/issues/2 to:
> Behavior that cleans up errors if implementationType is switched away from external

File: test/spec/provider/camunda-platform/CamundaPlatformPropertiesProvider.spec.js [hunk 5]
```diff
@@ -68,7 +68,7 @@ describe('<CamundaPlatformPropertiesProvider>', function() {
 
   describe('groups', function() {
 
-    describe('process', function() {
+    describe.only('process', function() {
```
Comment (MaxTru) [inline test/spec/provider/camunda-platform/CamundaPlatformPropertiesProvider.spec.js]:
```suggestion
    describe('process', function() {
```

**Comment by MaxTru** [comment]
The errors are currently alphabetically sorted by name (when re-opened). Should we not disable this sorting for errors, since the XML order has a meaning (e.g., the first error which matches the condition, will throw, but not the second one)?

**Comment by pinussilvestrus** [comment]
> > The error is not cleaned up, when we switch away from Implementationtype==external. Is there already a behavior that does this / or do we have an issue to create such a behavior?
> 
> FYI: I couldn't find such a behavior and created a task in #2 to:
> 
> > Behavior that cleans up errors if implementationType is switched away from external

I think we have one: https://github.com/camunda/camunda-bpmn-js/blob/main/lib/camunda-platform/features/modeling/behavior/DeleteErrorEventDefinitionBehavior.js

**Comment by MaxTru** [comment]
> > > The error is not cleaned up, when we switch away from Implementationtype==external. Is there already a behavior that does this / or do we have an issue to create such a behavior?
> > 
> > 
> > FYI: I couldn't find such a behavior and created a task in #2 to:
> > > Behavior that cleans up errors if implementationType is switched away from external
> 
> I think we have one: https://github.com/camunda/camunda-bpmn-js/blob/main/lib/camunda-platform/features/modeling/behavior/DeleteErrorEventDefinitionBehavior.js

Ups - that is great.

Sorry for the confusion everyone :-)

**Review by MaxTru — CHANGES_REQUESTED** [review]
only thing missing: https://github.com/bpmn-io/bpmn-properties-panel/pull/108#discussion_r714278352

**Comment by andreasgeier** [comment]
UX feedback:

1. I miss the message input. Has it been forgotten?
2. I miss the code part in the list item title (that is already [mentioned above](https://github.com/bpmn-io/bpmn-properties-panel/pull/108#issuecomment-921730347))
3. If the order of errors matters, we cannot use the alphabetical sort here. But since this is a general topic and [under discussion in another issue](https://github.com/bpmn-io/properties-panel/issues/96), we might fix that outside of this PR?
2. I know that the form structure reflects 1:1 what we have in the old panel. However, it breaks the list concept/structure where the first field of a list item always defines the list item title. So, from a UX perspective, it makes sense to have the global error reference select as the first form element and the throw expression at the end (below the select-related inputs). I also assume it would be the logical order: first define/reference the element (the global error), then the condition (the throw expression). Would that be okay from the developer's perspective, too?
3. The empty state label of a newly created item should be <no reference> (angle brackets) in line with <none> and <empty> that we already use in other lists and selects.

**Comment by MaxTru** [comment]
> If the order of errors matters, we cannot use the alphabetical sort here. But since this is a general topic and under discussion in another issue, we might fix that outside of this PR?

Yep, let's please ignore that here, since we will resolve this for all cases based on this issue: https://github.com/bpmn-io/properties-panel/issues/96

**Comment by philippfromme** [comment]
@andreasgeier I've applied the changes you requested. The only issue is that now the first entry is the select, which cannot be auto-focused. Therefore, we can either auto-focus the expression input or no input.

XHxyKbzOW3

**Comment by philippfromme** [comment]
`<select>` elements will be auto-focusable: https://github.com/bpmn-io/properties-panel/pull/101

This PR now depends on https://github.com/bpmn-io/properties-panel/pull/101.

**Comment by philippfromme** [comment]
@MaxTru Awaiting your approval.

File: test/spec/BpmnPropertiesPanelRenderer.spec.js [hunk 6]
```diff
@@ -89,6 +89,12 @@ describe('<BpmnPropertiesPanelRenderer>', function() {
       ...options
     });
 
+    modeler.on('commandStack.changed', async () => {
+      const { xml } = await modeler.saveXML({ format: true });
+
+      console.log(xml);
```
Comment (MaxTru) [inline test/spec/BpmnPropertiesPanelRenderer.spec.js:95]: I think this was forgotten here :-)

Comment (philippfromme) [reply test/spec/BpmnPropertiesPanelRenderer.spec.js:95]: Argh!

Comment (philippfromme) [reply test/spec/BpmnPropertiesPanelRenderer.spec.js]: Removed. 👍🏻

**Review by MaxTru — APPROVED** [review]
:clap:

**Comment by pinussilvestrus** [comment]
✌️

**Pull request merged**